File: CloudFlow/src/ro/cloudSoft/cloudDoc/presentation/client/shared/model/nomenclator/NomenclatorAttributeModel.java
package ro.cloudSoft.cloudDoc.presentation.client.shared.model.nomenclator;

import java.util.List;

import ro.cloudSoft.cloudDoc.domain.nomenclator.NomenclatorAttribute.NomenclatorAttributeAutocompleteType;

public class NomenclatorAttributeModel {
	
	private Long id;
	private Long nomenclatorId;
	private String name;
	private String key;
	private Long uiOrder;
	private String type;
	private Long typeNomenclatorId;
	private boolean required;
	private String requiredCheckExpression;
	private String invisibleCheckExpression;
	private String defaultValue;
	private boolean readonlyOnAdd;
	private boolean readonlyOnEdit;
	private List<NomenclatorAttributeSelectionFilterModel> typeNomenclatorSelectionFilters;
	private String typeNomenclatorSelectionFiltersCustomClass;
	private String typeNomenclatorSelectionFiltersCustomClassAttributeKeys;	
	private String attributeKeyForAutocomplete;
	private NomenclatorAttributeAutocompleteType autocompleteType;
	private String nomenclatorAttributeKeyForAutocomplete;
	private Long completionSuggestionNomenclatorId;
	private String completionSuggestionNomenclatorAttributeKey;
	
	public Long getId() {
		return id;
	}
	public void setId(Long id) {
		this.id = id;
	}
	public Long getNomenclatorId() {
		return nomenclatorId;
	}
	public void setNomenclatorId(Long nomenclatorId) {
		this.nomenclatorId = nomenclatorId;
	}
	public String getName() {
		return name;
	}
	public void setName(String name) {
		this.name = name;
	}
	public String getKey() {
		return key;
	}
	public void setKey(String key) {
		this.key = key;
	}
	public Long getUiOrder() {
		return uiOrder;
	}
	public void setUiOrder(Long uiOrder) {
		this.uiOrder = uiOrder;
	}
	public String getType() {
		return type;
	}
	public void setType(String type) {
		this.type = type;
	}
	public Long getTypeNomenclatorId() {
		return typeNomenclatorId;
	}
	public void setTypeNomenclatorId(Long typeNomenclatorId) {
		this.typeNomenclatorId = typeNomenclatorId;
	}
	public boolean isRequired() {
		return required;
	}
	public void setRequired(boolean required) {
		this.required = required;
	}
	public String getDefaultValue() {
		return defaultValue;
	}
	public void setDefaultValue(String defaultValue) {
		this.defaultValue = defaultValue;
	}
	public boolean isReadonlyOnAdd() {
		return readonlyOnAdd;
	}
	public void setReadonlyOnAdd(boolean readonlyOnAdd) {
		this.readonlyOnAdd = readonlyOnAdd;
	}
	public boolean isReadonlyOnEdit() {
		return readonlyOnEdit;
	}
	public void setReadonlyOnEdit(boolean readonlyOnEdit) {
		this.readonlyOnEdit = readonlyOnEdit;
	}
	public List<NomenclatorAttributeSelectionFilterModel> getTypeNomenclatorSelectionFilters() {
		return typeNomenclatorSelectionFilters;
	}
	public void setTypeNomenclatorSelectionFilters(
			List<NomenclatorAttributeSelectionFilterModel> typeNomenclatorSelectionFilters) {
		this.typeNomenclatorSelectionFilters = typeNomenclatorSelectionFilters;
	}
	public String getTypeNomenclatorSelectionFiltersCustomClass() {
		return typeNomenclatorSelectionFiltersCustomClass;
	}
	public void setTypeNomenclatorSelectionFiltersCustomClass(String typeNomenclatorSelectionFiltersCustomClass) {
		this.typeNomenclatorSelectionFiltersCustomClass = typeNomenclatorSelectionFiltersCustomClass;
	}
	public String getTypeNomenclatorSelectionFiltersCustomClassAttributeKeys() {
		return typeNomenclatorSelectionFiltersCustomClassAttributeKeys;
	}
	public void setTypeNomenclatorSelectionFiltersCustomClassAttributeKeys(
			String typeNomenclatorSelectionFiltersCustomClassAttributeKeys) {
		this.typeNomenclatorSelectionFiltersCustomClassAttributeKeys = typeNomenclatorSelectionFiltersCustomClassAttributeKeys;
	}
	public String getAttributeKeyForAutocomplete() {
		return attributeKeyForAutocomplete;
	}
	public void setAttributeKeyForAutocomplete(String attributeKeyForAutocomplete) {
		this.attributeKeyForAutocomplete = attributeKeyForAutocomplete;
	}
	public NomenclatorAttributeAutocompleteType getAutocompleteType() {
		return autocompleteType;
	}
	public void setAutocompleteType(NomenclatorAttributeAutocompleteType autocompleteType) {
		this.autocompleteType = autocompleteType;
	}
	public String getNomenclatorAttributeKeyForAutocomplete() {
		return nomenclatorAttributeKeyForAutocomplete;
	}
	public void setNomenclatorAttributeKeyForAutocomplete(String nomenclatorAttributeKeyForAutocomplete) {
		this.nomenclatorAttributeKeyForAutocomplete = nomenclatorAttributeKeyForAutocomplete;
	}
	public String getRequiredCheckExpression() {
		return requiredCheckExpression;
	}
	public void setRequiredCheckExpression(String requiredCheckExpression) {
		this.requiredCheckExpression = requiredCheckExpression;
	}
	public String getInvisibleCheckExpression() {
		return invisibleCheckExpression;
	}
	public void setInvisibleCheckExpression(String invisibleCheckExpression) {
		this.invisibleCheckExpression = invisibleCheckExpression;
	}
	public Long getCompletionSuggestionNomenclatorId() {
		return completionSuggestionNomenclatorId;
	}
	public void setCompletionSuggestionNomenclatorId(Long completionSuggestionNomenclatorId) {
		this.completionSuggestionNomenclatorId = completionSuggestionNomenclatorId;
	}
	public String getCompletionSuggestionNomenclatorAttributeKey() {
		return completionSuggestionNomenclatorAttributeKey;
	}
	public void setCompletionSuggestionNomenclatorAttributeKey(String completionSuggestionNomenclatorAttributeKey) {
		this.completionSuggestionNomenclatorAttributeKey = completionSuggestionNomenclatorAttributeKey;
	}
}

File: CloudFlow/FrontEnd-App/src/app/shared/model/nomenclator/nomenclator-attribute.model.ts
import { JsonObject, JsonProperty } from "json2typescript";
import { NomenclatorAttributeSelectionFilterModel } from "./nomenclator-attribute-selection-filter.model";
import { ArrayUtils } from "./../../utils";

@JsonObject
export class NomenclatorAttributeModel {
	
	public static readonly AUTOCOMPLETE_TYPE_NOMENCLATOR_ATTRIBUTE: string = "NOMENCLATOR_ATTRIBUTE";
	public static readonly AUTOCOMPLETE_TYPE_COPY_VALUE: string = "COPY_VALUE";
	public static readonly AUTOCOMPLETE_TYPE_DELETE_VALUE: string = "DELETE_VALUE";

	@JsonProperty("id", Number)
	public id: number = null;

	@JsonProperty("nomenclatorId", Number)
	public nomenclatorId: number = null;

	@JsonProperty("name", String)
	public name: string = null;

	@JsonProperty("key", String)
	public key: string = null;

	@JsonProperty("uiOrder", Number)
	public uiOrder: number = null;

	@JsonProperty("type", String)
	public type: string = null;

	@JsonProperty("typeNomenclatorId", Number)
	public typeNomenclatorId: number = null;

	@JsonProperty("typeNomenclatorSelectionFilters", [NomenclatorAttributeSelectionFilterModel], true)
	public typeNomenclatorSelectionFilters: NomenclatorAttributeSelectionFilterModel[] = [];

	@JsonProperty("typeNomenclatorSelectionFiltersCustomClass", String, true)
	public typeNomenclatorSelectionFiltersCustomClass: string = null;

	@JsonProperty("typeNomenclatorSelectionFiltersCustomClassAttributeKeys", String, true)
	public typeNomenclatorSelectionFiltersCustomClassAttributeKeys: string = null;
	
	@JsonProperty("attributeKeyForAutocomplete", String, true)
	public attributeKeyForAutocomplete: string = null;

	@JsonProperty("autocompleteType", String, true)
	public autocompleteType: string = undefined;

	@JsonProperty("nomenclatorAttributeKeyForAutocomplete", String, true)
	public nomenclatorAttributeKeyForAutocomplete: string = null;

	@JsonProperty("required", Boolean)
	public required: boolean = false;

	@JsonProperty("requiredCheckExpression", String, true)
	public requiredCheckExpression: string = null;

	@JsonProperty("invisibleCheckExpression", String, true)
	public invisibleCheckExpression: string = null;

	@JsonProperty("defaultValue", String, true)
	public defaultValue: string = null;

	@JsonProperty("readonlyOnAdd", Boolean, true)
	public readonlyOnAdd: boolean = false;

	@JsonProperty("readonlyOnEdit", Boolean, true)
	public readonlyOnEdit: boolean = false;

	@JsonProperty("completionSuggestionNomenclatorId", Number, true)
	public completionSuggestionNomenclatorId: number = null;

	@JsonProperty("completionSuggestionNomenclatorAttributeKey", String, true)
	public completionSuggestionNomenclatorAttributeKey: string = null;

	// TODO - De adaugata si celelalte proprietati cand va fi nevoie

	public clone(): NomenclatorAttributeModel {
		let clone: NomenclatorAttributeModel = new NomenclatorAttributeModel();
		clone.id = this.id;
		clone.name = this.name;
		clone.nomenclatorId = this.nomenclatorId;
		clone.key = this.key;
		clone.uiOrder = this.uiOrder;
		clone.type = this.type;
		clone.typeNomenclatorId = this.typeNomenclatorId;
		clone.typeNomenclatorSelectionFilters = [];
		if (ArrayUtils.isNotEmpty(this.typeNomenclatorSelectionFilters)) {
			this.typeNomenclatorSelectionFilters.forEach((filter: NomenclatorAttributeSelectionFilterModel) => {
				clone.typeNomenclatorSelectionFilters.push(filter.clone());
			});
		}
		clone.typeNomenclatorSelectionFiltersCustomClass = this.typeNomenclatorSelectionFiltersCustomClass;
		clone.typeNomenclatorSelectionFiltersCustomClassAttributeKeys = this.typeNomenclatorSelectionFiltersCustomClassAttributeKeys;
		clone.defaultValue = this.defaultValue;
		clone.required = this.required;
		clone.readonlyOnAdd = this.readonlyOnAdd;
		clone.readonlyOnEdit = this.readonlyOnEdit;
		clone.attributeKeyForAutocomplete = this.attributeKeyForAutocomplete;
		clone.autocompleteType = this.autocompleteType;
		clone.nomenclatorAttributeKeyForAutocomplete = this.nomenclatorAttributeKeyForAutocomplete;
		clone.completionSuggestionNomenclatorId = this.completionSuggestionNomenclatorId;
		clone.completionSuggestionNomenclatorAttributeKey = this.completionSuggestionNomenclatorAttributeKey;
		return clone;
	}
}

export enum AttributeTypeEnum {
	
	TEXT = "TEXT",
	NUMERIC = "NUMERIC",
	DATE = "DATE",
	BOOLEAN = "BOOLEAN",
	NOMENCLATOR = "NOMENCLATOR",
}